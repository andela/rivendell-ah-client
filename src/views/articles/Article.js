import React from 'react';
const Article = () =>(
<div>
        <h2>A user Article</h2>
    </div>
);
    export default Article;