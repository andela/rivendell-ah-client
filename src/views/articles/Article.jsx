import React from 'react';
import { Link } from 'react-router-dom';
import { PropTypes } from 'prop-types';
import { Item, Button, Icon } from 'semantic-ui-react';
import { connect } from 'react-redux';
import articleAction from '../../actions/articleAction';
import likeAction from '../../actions/likeAction';
import ArticleView from '../../components/ArticleView';
/**
 * Article Component
 */
export class Article extends React.Component {
  /**
   * constructor function
   */
  constructor() {
    super();
    this.state = {
      likeAttributes: {
        divClass: 'ui labeled button',
        buttonClass: 'ui button',
        iconClass: 'heart icon',
        linkClass: 'ui basic left pointing label'
      },
      start: 0,
      end: 2,
      range: 2,
      showPreviousIcon: false,
      disableShowMore: false,
    };
    this.handleClick = this.handleClick.bind(this);
    this.checkUserLikes = this.checkUserLikes.bind(this);
    this.onLoadMore = this.onLoadMore.bind(this);
    this.renderLikeBy = this.renderLikeBy.bind(this);
    this.onShowPrevious = this.onShowPrevious.bind(this);
  }

  /**
  * ComponentDidMount lifecycle
  * @returns {string} - HTML Markup for the component
  */
  componentDidMount() {
    const { match, getArticle } = this.props;
    const { params } = match;
    // dispatch an article (an example just to test the like functionality)
    getArticle(params.slug);
  }

  /**
  * componentWillReceiveProps lifecycle
  * @returns {string} - HTML Markup for the component
  * @param {object} nextProps
  */
  componentWillReceiveProps(nextProps) {
    if (nextProps.like
      || this.checkUserLikes(nextProps.likes)) {
      this.setState({
        likeAttributes: {
          divClass: 'ui labeled button',
          buttonClass: 'ui blue button',
          iconClass: 'heart icon',
          linkClass: 'ui basic left pointing label'
        }
      });
    } else {
      this.setState({
        likeAttributes: {
          divClass: 'ui labeled button',
          buttonClass: 'ui button',
          iconClass: 'heart icon',
          linkClass: 'ui basic left pointing label'
        }
      });
    }
  }


  /**
  * ComponentDidUpdate lifecycle
  * @returns {string} - HTML Markup for the component
  * @param {object} prevProps
  * @param {object} prevState
  */
  componentDidUpdate(prevProps) {
    const { article, getLikes, like, match } = this.props;
    const { params } = match;
    if (article.likes !== prevProps.article.likes
      || like !== prevProps.like) {
      getLikes(params.slug);
    }
  }


  /**
  * onLoadmore
  * @returns {null} - Returns null
  */
  onLoadMore() {
    const { likesCount } = this.props;
    let { start, end,
      showPreviousIcon, disableShowMore
    } = this.state;
    const { range } = this.state;
    end += 1;
    if (end - start > range) start += 1;
    if (end >= likesCount) {
      end = likesCount;
      disableShowMore = true;
    }
    if (start > 0) {
      showPreviousIcon = true;
    }
    this.setState({
      start,
      end,
      showPreviousIcon,
      disableShowMore
    });
  }

  /**
 * onShowPrevious
 * @returns {null} - Returns null
 */
  onShowPrevious() {
    const { likesCount } = this.props;
    let { start, end,
      showPreviousIcon, disableShowMore
    } = this.state;
    const { range } = this.state;
    if (start === 0) return;
    start -= 1;
    if (end - start > range) end -= 1;
    if (start <= 0) {
      start = 0;
      showPreviousIcon = false;
    }
    if (end < likesCount) disableShowMore = false;
    this.setState({
      start,
      end,
      showPreviousIcon,
      disableShowMore
    });
  }

  /**
    * checkUserLikes
    * @returns {boolean} - HTML Markup for the component
    * @param {Array} likes
    */
  checkUserLikes(likes) {
    const { userProfile } = this.props;
    return likes.some(like => like.user.id === userProfile.id);
  }

  /**
   * handle form submit
   * @param {Object} event event object
   * @returns {null} null
   */
  handleClick() {
    const {
      match, token, history,
      likeArticle, unlikeArticle,
      like, likes
    } = this.props;
    const { params } = match;
    if (token) {
      if (!like && !this.checkUserLikes(likes)) {
        likeArticle(params.slug, token);
      } else {
        unlikeArticle(params.slug, token);
      }
    } else {
      history.push('/login');
    }
  }

  /**
  * renderLikeby
  * @returns {jsx} - HTML Markup for the component
  */
  renderLikeBy() {
    const { likes } = this.props;
    const { start, end, disableShowMore, showPreviousIcon } = this.state;
    const list = likes.length > 0
      ? likes.slice(start, end).map((likedBy, index) => (
        <Item key={index}>
          {likedBy.user.image ? (
            <Item.Image
              size="tiny"
              src={likedBy.user.image}
            />
          )
            : (
              <Icon
                size="massive"
                name="user"
                color="blue"
              />
            )}

          <Item.Content>
            <Item.Header>
              <Link to={`/@${likedBy.user.username}`}>
                {`${likedBy.user.firstName} 
              ${likedBy.user.lastName}`}
              </Link>
            </Item.Header>
            <Item.Meta>Bio</Item.Meta>
            <Item.Description>
              {likedBy.user.bio
                ? likedBy.user.bio : 'No Bio Info'}
            </Item.Description>
            <Item.Extra>
              {`Liked: ${likedBy.timeLiked.substring(0, 10)}`}
            </Item.Extra>
          </Item.Content>
        </Item>
      )) : ' ';
    return typeof list === 'object'
      ? (
        <div>
          <Item.Group>
            <div
              className={showPreviousIcon
                ? 'show-more' : 'hide'}
            >
              <Icon
                className="pointer"
                color="blue"
                name="chevron up"
                size="large"
                onClick={this.onShowPrevious}
              />
            </div>
            <h5 className="likes-header">
              Users Who Liked
            </h5>
            {list}
            <div
              className="show-more"
            >
              <Button
                color="blue"
                className="ui button"
                onClick={this.onLoadMore}
                disabled={disableShowMore}
                basic
              >
                Show More...
              </Button>
            </div>

          </Item.Group>
        </div>
      ) : 'No Likes';
  }

  /**
 * render function
 * @returns {Function} jsx
 */
  render() {
    const { likeAttributes } = this.state;
    const { likesCount, likes, match: { params } } = this.props;
    const likeProps = {
      likeAttributes,
      likesCount,
      handleClick: this.handleClick,
      likes,
      renderLikeBy: this.renderLikeBy,
      onLoadMore: this.onLoadMore,
    };
    return (
      <div>
        <ArticleView slug={params.slug} likeProps={likeProps} />
      </div>
    );
  }
}
export const mapStateToProps = (state) => {
  const { article } = state.article;
  const { token } = state.auth;
  const { userProfile } = state.profile;
  const { likes, like, likesCount } = state.like;
  return {
    article,
    userProfile,
    likes,
    like,
    likesCount,
    token,
  };
};
Article.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.object,
  }).isRequired,
  likesCount: PropTypes.number,
  getArticle: PropTypes.func.isRequired,
  getLikes: PropTypes.func,
  like: PropTypes.bool.isRequired,
  userProfile: PropTypes.shape({
    id: PropTypes.string,
  }),
  likes: PropTypes.arrayOf(PropTypes.shape({
    user: PropTypes.object
  })),
  article: PropTypes.shape({
    likes: PropTypes.array,
  }),
  token: PropTypes.string,
  history: PropTypes.shape({
    push: PropTypes.func
  }),
  likeArticle: PropTypes.func,
  unlikeArticle: PropTypes.func,

};
Article.defaultProps = {
  likesCount: 0,
  getLikes: () => { },
  userProfile: {},
  likes: [],
  article: PropTypes.shape({
    likes: [],
  }),
  token: '',
  history: {},
  likeArticle: () => { },
  unlikeArticle: () => { },
};
export default connect(mapStateToProps, {
  getArticle: articleAction.getArticle,
  getLikes: likeAction.getLikes,
  likeArticle: likeAction.likeArticle,
  unlikeArticle: likeAction.unlikeArticle,
})(Article);



// import { Label } from 'semantic-ui-react';

// const Article = () => (
//   <div>
//     <h2>A User Article</h2>
//     <Label.Group tag>
//       <Label as="a">
//         <Link to={`/article-tags/andela`}>andela</Link>
//       </Label>
//       <Label as="a">fatti</Label>
//       <Label as="a">love</Label>
//       <Label as="a">issue</Label>
//       <Label as="a">cup</Label>
//     </Label.Group>
//   </div>
// );
// export default Article;
