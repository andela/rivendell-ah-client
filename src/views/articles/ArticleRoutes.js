import React from 'react';
import { Switch, Route } from 'react-router-dom';
import articlesPages from '../articles'; 
import authPages from '../auth';
const { AuthRoute } = authPages;

const { Article, UserArticles, EditArticle, NewArticle } = articlesPages;

const UserArticle = () => (
  <div>
    <Switch>
      <Route exact path='/articles' component={UserArticles} /> 
      <AuthRoute exact path='/articles/new' component={NewArticle} />
      <Route exact path='/articles/:slug' component={Article} />
      <AuthRoute exact path='/articles/:slug/edit' component={EditArticle} />
    </Switch>
  </div>
);

export default UserArticle;