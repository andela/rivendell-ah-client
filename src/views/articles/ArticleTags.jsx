import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button, Item, Label } from 'semantic-ui-react';
import { PropTypes } from 'prop-types';
import getArticlesByTags from '../../actions/tagAction';
import ListArticlesByTag from './ListArticlesByTag';

/**
 * ArticleTags
 * @returns {Null} jsx
 *
 */
export class ArticleTags extends React.Component {
  /**
   * constructor function
   * @param {Object} props props object
   */
  constructor() {
    super();
    this.state = {};
  }

  /**
   * component did mount
   * @returns {null} null
   */
  componentDidMount() {
    const { dispatch, match } = this.props;
    const { tag } = match.params;
    dispatch(getArticlesByTags(tag));
  }

  /**
   * @returns {Function} jsx
   */
  render() {
    const { articles } = this.props;
    const content = articles.map((article, index) => (
      <Item key={index}>
        <Item.Image
          src={article.image
          || 'https://react.semantic-ui.com/images/wireframe/image.png'}
        />

        <Item.Content>
          <Item.Header as="a">{article.title}</Item.Header>
          <Item.Meta>
            <span className="cinema">{article.description}</span>
          </Item.Meta>
          <Item.Description>
            {`${article.body.slice(0, 150)}...`}
          </Item.Description>
          <Link to={`/article/${article.slug}`}>
            <Button className="read-more" content="Read more..." />
          </Link>
          <Item.Extra>
            {article.tags.map(tag => (
              <Link to={`/article-tags/${tag}`}><Label>{tag}</Label></Link>
            ))}
          </Item.Extra>
        </Item.Content>
      </Item>
    ));
    return (
      <div className="central">
        <ListArticlesByTag content={content} />
        {/* <pre>{JSON.stringify(articles, null, 2)}</pre> */}
      </div>
    );
  }
}

ArticleTags.propTypes = {
  dispatch: PropTypes.bool.isRequired,
  match: PropTypes.shape({}).isRequired,
  errors: PropTypes.shape({}).isRequired,
  articles: PropTypes.shape({}).isRequired,
};

export const mapStateToProps = (state) => {
  const { errors, articles, isLoading } = state.articlesByTags;
  return {
    errors,
    articles,
    isLoading
  };
};

export default connect(mapStateToProps)(ArticleTags);
