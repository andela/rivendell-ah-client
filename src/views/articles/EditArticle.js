import React from 'react';
const EditArticle = () => (
    <div>
        <h2>Edit an Article</h2>
    </div>
);
export default EditArticle;