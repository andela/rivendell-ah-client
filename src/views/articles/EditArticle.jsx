import React, { Component } from 'react';
import { Container } from 'semantic-ui-react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import EditArticleControls from '../../components/EditArticleControls';
import updateArticleAction from '../../actions/updateArticleAction';
import loadArticleAction from '../../actions/loadArticleAction';


/**
 * This component manages the state of the input elements
 * to the edit Articles
 * @class NewArticle
 * @returns {object} instance of an object
 */
export class EditArticle extends Component {
  /**
   * creates an instance of a NewArticle and binds the
   * methods to the this
   * @param {object} props contains the props sent from thd parent
   */
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
  }


  /**
   *called once this component has mounted successfully to
   the dom
   @returns {void}  performs an action and returns nothning
   */
  componentDidMount() {
    const { loadArticle, match: { params } } = this.props;
    loadArticle(params.slug);
  }


  /**
   *
   * This function is called when the the user
   * tries to submit the article
   * @returns {void} performs an action and returns nothing
   * @param {object} article contains information about the
   * article
   */
  handleSubmit(article) {
    const { updateArticle, match: { params } } = this.props;
    const { slug } = params;
    updateArticle(article, slug);
  }

  /**
   * this renders html
   * @returns {void} renders html an returns nothing
   */
  render() {
    const { storeState } = this.props;
    return (
      <Container>
        <EditArticleControls
          onSubmit={this.handleSubmit}
          {...storeState}
          update
        />
      </Container>);
  }
}


EditArticle.propTypes = {
  updateArticle: propTypes.func,
  loadArticle: propTypes.func,
  match: propTypes.objectOf(propTypes.object),
  storeState: propTypes.shape({
    editing: propTypes.bool,
    isLoading: propTypes.bool,
    success: propTypes.bool,
    article: propTypes.object,
    errors: propTypes.shape({
      description: propTypes.arrayOf(propTypes.string),
      title: propTypes.arrayOf(propTypes.string),
    }),
  })
};

EditArticle.defaultProps = {
  updateArticle: () => {},
  loadArticle: () => {},
  match: { params: { slug: '' } },
  storeState: {},
};

export const mapDispatchToProps = dispatch => ({
  updateArticle:
    (article, slug) => dispatch(updateArticleAction(article, slug)),
  loadArticle: slug => dispatch(loadArticleAction(slug)),
});

export const mapStateToProps = state => ({
  storeState: state.updateArticleReducer,
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditArticle);
