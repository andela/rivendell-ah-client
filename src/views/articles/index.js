import Article from './Article';
import UserArticles from './UserArticles';
import EditArticle from './EditArticle';
import NewArticle from './NewArticle';

export default {
  Article,
  UserArticles,
  EditArticle,
  NewArticle,
};
