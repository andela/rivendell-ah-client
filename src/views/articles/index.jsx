import UserArticles from './UserArticles';
import EditArticle from './EditArticle';
import NewArticle from './NewArticle';
import Article from './Article';
import ArticleTags from './ArticleTags';

export default {
  UserArticles,
  EditArticle,
  NewArticle,
  Article,
  ArticleTags,
};
