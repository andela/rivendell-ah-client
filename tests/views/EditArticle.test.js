import { shallow } from 'enzyme';
import React from 'react';
import chai from 'chai';
import { EditArticle, mapStateToProps, mapDispatchToProps } from '../../src/views/articles/EditArticle';

describe('testing NewArticle Copmonent', () => {
  describe('basic components it renders', () => {
    it('should render an EditArticleControls component ', () => {
      const mountedObj = shallow(<EditArticle />);
      expect(mountedObj)
        .toMatchSnapshot();
    });
  });


  describe('functions in the comonent', () => {
    it('should call persistArticle with its argument', () => {
      const updateArticleSpy = jest.fn();
      const mockSlug = { slug: 'mockSlug' };
      const mountedObj = shallow(
        <EditArticle
          updateArticle={updateArticleSpy}
          match={{ params: mockSlug }}
        />
      );
      const expectedArg = 'expect this argumet';
      mountedObj.instance().handleSubmit(expectedArg);
      expect(updateArticleSpy)
        .toHaveBeenCalledWith(expectedArg, mockSlug.slug);
    });
  });

  describe('mapStateToProps', () => {
    it('mapStateToProps should correctly map the state passed as argument', () => {
      const state = {
        updateArticleReducer: 'Hello code'
      };
      const mappedState = mapStateToProps(state);

      chai.expect(mappedState.storeState)
        .to.equal(state.updateArticleReducer);
    });
  });

  describe('mapDispatchToProps', () => {
    it('mapDispatchToProps should correctly map the state passed as argument', () => {
      const dispatchSpy = jest.fn();
      const mappedState = mapDispatchToProps(dispatchSpy);

      chai.expect(mappedState.updateArticle)
        .to.be.a('function');
      const expectArg = 'this argument';
      mappedState.updateArticle(expectArg);
      expect(dispatchSpy)
        .toHaveBeenCalledTimes(1);
    });

    it('mapDispatchToProps should correctly map the state passed as argument', () => {
      const dispatchSpy = jest.fn();
      const mappedState = mapDispatchToProps(dispatchSpy);

      chai.expect(mappedState.loadArticle)
        .to.be.a('function');
      const expectArg = 'this argument';
      mappedState.loadArticle(expectArg);
      expect(dispatchSpy)
        .toHaveBeenCalledTimes(1);
    });
  });

  describe('defaultProps', () => {
    it('should have a persist article that is an empty functuon', () => {
      chai.expect(EditArticle.defaultProps.updateArticle)
        .to.be.a('function');
      chai.expect(EditArticle.defaultProps.updateArticle())
        .to.equal(undefined);
    });
  });
});
