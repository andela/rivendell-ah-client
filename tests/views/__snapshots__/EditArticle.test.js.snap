// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing NewArticle Copmonent basic components it renders should render an EditArticleControls component  1`] = `
<Container>
  <Connect(EditArticleControls)
    onSubmit={[Function]}
    update={true}
  />
</Container>
`;
